diff --git a/include/locale.h b/include/locale.h
index ce384381..d0c26c9f 100644
--- a/include/locale.h
+++ b/include/locale.h
@@ -7,11 +7,13 @@ extern "C" {
 
 #include <features.h>
 
+#ifndef NULL
 #ifdef __cplusplus
 #define NULL 0L
 #else
 #define NULL ((void*)0)
 #endif
+#endif
 
 #define LC_CTYPE    0
 #define LC_NUMERIC  1
diff --git a/include/math.h b/include/math.h
index 14f28ec8..bcb19851 100644
--- a/include/math.h
+++ b/include/math.h
@@ -65,39 +65,18 @@ static __inline unsigned long long __DOUBLE_BITS(double __f)
 	return __u.__i;
 }
 
-#define fpclassify(x) ( \
-	sizeof(x) == sizeof(float) ? __fpclassifyf(x) : \
-	sizeof(x) == sizeof(double) ? __fpclassify(x) : \
-	__fpclassifyl(x) )
-
-#define isinf(x) ( \
-	sizeof(x) == sizeof(float) ? (__FLOAT_BITS(x) & 0x7fffffff) == 0x7f800000 : \
-	sizeof(x) == sizeof(double) ? (__DOUBLE_BITS(x) & -1ULL>>1) == 0x7ffULL<<52 : \
-	__fpclassifyl(x) == FP_INFINITE)
-
-#define isnan(x) ( \
-	sizeof(x) == sizeof(float) ? (__FLOAT_BITS(x) & 0x7fffffff) > 0x7f800000 : \
-	sizeof(x) == sizeof(double) ? (__DOUBLE_BITS(x) & -1ULL>>1) > 0x7ffULL<<52 : \
-	__fpclassifyl(x) == FP_NAN)
-
-#define isnormal(x) ( \
-	sizeof(x) == sizeof(float) ? ((__FLOAT_BITS(x)+0x00800000) & 0x7fffffff) >= 0x01000000 : \
-	sizeof(x) == sizeof(double) ? ((__DOUBLE_BITS(x)+(1ULL<<52)) & -1ULL>>1) >= 1ULL<<53 : \
-	__fpclassifyl(x) == FP_NORMAL)
-
-#define isfinite(x) ( \
-	sizeof(x) == sizeof(float) ? (__FLOAT_BITS(x) & 0x7fffffff) < 0x7f800000 : \
-	sizeof(x) == sizeof(double) ? (__DOUBLE_BITS(x) & -1ULL>>1) < 0x7ffULL<<52 : \
-	__fpclassifyl(x) > FP_INFINITE)
+int fpclassify(double);
+
+int isfinite(double x);
+int isinf(double x);
+int isnan(double x);
+int isnormal(double x);
 
 int __signbit(double);
 int __signbitf(float);
 int __signbitl(long double);
 
-#define signbit(x) ( \
-	sizeof(x) == sizeof(float) ? (int)(__FLOAT_BITS(x)>>31) : \
-	sizeof(x) == sizeof(double) ? (int)(__DOUBLE_BITS(x)>>63) : \
-	__signbitl(x) )
+int signbit(double);
 
 #define isunordered(x,y) (isnan((x)) ? ((void)(y),1) : isnan((y)))
 
@@ -200,6 +179,7 @@ double      expm1(double);
 float       expm1f(float);
 long double expm1l(long double);
 
+int         abs(int);
 double      fabs(double);
 float       fabsf(float);
 long double fabsl(long double);
diff --git a/include/stddef.h b/include/stddef.h
index bd753853..9f20990a 100644
--- a/include/stddef.h
+++ b/include/stddef.h
@@ -1,11 +1,13 @@
 #ifndef _STDDEF_H
 #define _STDDEF_H
 
+#ifndef NULL
 #ifdef __cplusplus
 #define NULL 0L
 #else
 #define NULL ((void*)0)
 #endif
+#endif
 
 #define __NEED_ptrdiff_t
 #define __NEED_size_t
@@ -16,10 +18,12 @@
 
 #include <bits/alltypes.h>
 
+#ifndef offsetof
 #if __GNUC__ > 3
 #define offsetof(type, member) __builtin_offsetof(type, member)
 #else
 #define offsetof(type, member) ((size_t)( (char *)&(((type *)0)->member) - (char *)0 ))
 #endif
+#endif
 
 #endif
diff --git a/include/stdio.h b/include/stdio.h
index 3604198c..ba195062 100644
--- a/include/stdio.h
+++ b/include/stdio.h
@@ -25,11 +25,13 @@ extern "C" {
 
 #include <bits/alltypes.h>
 
+#ifndef NULL
 #ifdef __cplusplus
 #define NULL 0L
 #else
 #define NULL ((void*)0)
 #endif
+#endif
 
 #undef EOF
 #define EOF (-1)
diff --git a/include/stdlib.h b/include/stdlib.h
index b54a051f..efb3b0a1 100644
--- a/include/stdlib.h
+++ b/include/stdlib.h
@@ -7,11 +7,13 @@ extern "C" {
 
 #include <features.h>
 
+#ifndef NULL
 #ifdef __cplusplus
 #define NULL 0L
 #else
 #define NULL ((void*)0)
 #endif
+#endif
 
 #define __NEED_size_t
 #define __NEED_wchar_t
diff --git a/include/string.h b/include/string.h
index 795a2abc..df0b98b9 100644
--- a/include/string.h
+++ b/include/string.h
@@ -7,11 +7,13 @@ extern "C" {
 
 #include <features.h>
 
+#ifndef NULL
 #ifdef __cplusplus
 #define NULL 0L
 #else
 #define NULL ((void*)0)
 #endif
+#endif
 
 #define __NEED_size_t
 #if defined(_POSIX_SOURCE) || defined(_POSIX_C_SOURCE) \
diff --git a/include/time.h b/include/time.h
index 5494df18..1b180245 100644
--- a/include/time.h
+++ b/include/time.h
@@ -7,11 +7,13 @@ extern "C" {
 
 #include <features.h>
 
+#ifndef NULL
 #ifdef __cplusplus
 #define NULL 0L
 #else
 #define NULL ((void*)0)
 #endif
+#endif
 
 
 #define __NEED_size_t
diff --git a/include/wchar.h b/include/wchar.h
index 88eb55b1..9d03ca25 100644
--- a/include/wchar.h
+++ b/include/wchar.h
@@ -38,11 +38,13 @@ extern "C" {
 #define WCHAR_MIN (-1-0x7fffffff+L'\0')
 #endif
 
+#ifndef NULL
 #ifdef __cplusplus
 #define NULL 0L
 #else
 #define NULL ((void*)0)
 #endif
+#endif
 
 #undef WEOF
 #define WEOF 0xffffffffU
diff --git a/src/errno/__errno_location.c b/src/errno/__errno_location.c
index 7f9d6027..17d13563 100644
--- a/src/errno/__errno_location.c
+++ b/src/errno/__errno_location.c
@@ -6,4 +6,6 @@ int *__errno_location(void)
 	return &__pthread_self()->errno_val;
 }
 
-weak_alias(__errno_location, ___errno_location);
+int *___errno_location(void) {
+	return &__pthread_self()->errno_val;
+}
diff --git a/src/errno/strerror.c b/src/errno/strerror.c
index 7f926432..4b081389 100644
--- a/src/errno/strerror.c
+++ b/src/errno/strerror.c
@@ -39,8 +39,7 @@ char *__strerror_l(int e, locale_t loc)
 	return (char *)LCTRANS(s, LC_MESSAGES, loc);
 }
 
-char *strerror(int e)
-{
+char *strerror(int e) {
 	return __strerror_l(e, CURRENT_LOCALE);
 }
 
diff --git a/src/math/__fpclassify.c b/src/math/__fpclassify.c
index f7c0e2df..a672f2db 100644
--- a/src/math/__fpclassify.c
+++ b/src/math/__fpclassify.c
@@ -9,3 +9,27 @@ int __fpclassify(double x)
 	if (e==0x7ff) return u.i<<12 ? FP_NAN : FP_INFINITE;
 	return FP_NORMAL;
 }
+
+int fpclassify(double x) {
+	return __fpclassify(x);
+}
+
+int signbit(double x) {
+	return __signbit(x);
+}
+
+int isfinite(double x) {
+	return (__DOUBLE_BITS(x) & -1ULL>>1) < 0x7ffULL<<52;
+}
+
+int isinf(double x) {
+	return (__DOUBLE_BITS(x) & -1ULL>>1) == 0x7ffULL<<52;
+}
+
+int isnan(double x) {
+	return (__DOUBLE_BITS(x) & -1ULL>>1) > 0x7ffULL<<52;
+}
+
+int isnormal(double x) {
+	return ((__DOUBLE_BITS(x)+(1ULL<<52)) & -1ULL>>1) >= 1ULL<<53;
+}
diff --git a/src/stdio/printf.c b/src/stdio/printf.c
index cebfe404..e69de29b 100644
--- a/src/stdio/printf.c
+++ b/src/stdio/printf.c
@@ -1,12 +0,0 @@
-#include <stdio.h>
-#include <stdarg.h>
-
-int printf(const char *restrict fmt, ...)
-{
-	int ret;
-	va_list ap;
-	va_start(ap, fmt);
-	ret = vfprintf(stdout, fmt, ap);
-	va_end(ap);
-	return ret;
-}
diff --git a/src/string/x86_64/memcpy.s b/src/string/x86_64/memcpy.s
index 3d960efa..aa495768 100644
--- a/src/string/x86_64/memcpy.s
+++ b/src/string/x86_64/memcpy.s
@@ -1,3 +1,4 @@
+//*
 .global memcpy
 .global __memcpy_fwd
 .hidden __memcpy_fwd
@@ -23,3 +24,4 @@ __memcpy_fwd:
 	dec %edx
 	jnz 2b
 1:	ret
+//*/
